### ISSUE TITLE
Problem with updating database from 4.3.0 to 4.3.4 - running in docker

### ISSUE THREAD

**Issue by clix10** [LycheeOrg/Lychee#1057]
Hello guys, 

I was updating some docker images today, one of them was lychee. I am running it in docker. I updated the "lychee" and "lychee_db" to the latest version. 

After that I was greeted with the screen, that migration is required and I should enter admin username and password. If I do that, the next screen shows: "500 | SERVER ERROR"

This is start up log:

```
lychee_db    | 2021-07-11 13:04:22+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.5.11+maria~focal started.
lychee_db    | 2021-07-11 13:04:23+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
lychee_db    | 2021-07-11 13:04:23+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.5.11+maria~focal started.
lychee_db    | 2021-07-11 13:04:23 0 [Note] mysqld (mysqld 10.5.11-MariaDB-1:10.5.11+maria~focal) starting as process 1 ...
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Uses event mutexes
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Number of pools: 1
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Using crc32 + pclmulqdq instructions
lychee_db    | 2021-07-11 13:04:23 0 [Note] mysqld: O_TMPFILE is not supported on /tmp (disabling future attempts)
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Using Linux native AIO
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Initializing buffer pool, total size = 134217728, chunk size = 134217728
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Completed initialization of buffer pool
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: 128 rollback segments are active.
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Creating shared tablespace for temporary tables
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: 10.5.11 started; log sequence number 176961008; transaction id 538780
lychee_db    | 2021-07-11 13:04:23 0 [Note] Plugin 'FEEDBACK' is disabled.
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
lychee_db    | 2021-07-11 13:04:23 0 [Note] Server socket created on IP: '::'.
lychee_db    | 2021-07-11 13:04:23 0 [Warning] 'proxies_priv' entry '@% root@eedbed19ee44' ignored in --skip-name-resolve mode.
lychee       |
lychee       | -------------------------------------
lychee       |   _               _
lychee       |  | |   _   _  ___| |__   ___  ___
lychee       |  | |  | | | |/ __|  _ \ / _ \/ _ \
lychee       |  | |__| |_| | (__| | | |  __/  __/
lychee       |  |_____\__, |\___|_| |_|\___|\___|
lychee       |  | |   |___/ _ __ __ ___   _____| |
lychee       |  | |   / _' | '__/ _' \ \ / / _ \ |
lychee       |  | |__| (_| | | | (_| |\ V /  __/ |
lychee       |  |_____\__,_|_|  \__,_| \_/ \___|_|
lychee       |
lychee       | -------------------------------------
lychee       | Lychee Version: 4.3.4 (release)
lychee       | Lychee Commit:  8d41c0f
lychee       | https://github.com/LycheeOrg/Lychee/commit/8d41c0f928b3ffac8f07523a6b369f3daeee28df
lychee       | -------------------------------------
lychee       | **** Delaying startup (30 seconds)... ****
lychee_db    | 2021-07-11 13:04:23 0 [Note] Reading of all Master_info entries succeeded
lychee_db    | 2021-07-11 13:04:23 0 [Note] Added new Master_info '' to hash table
lychee_db    | 2021-07-11 13:04:23 0 [Note] mysqld: ready for connections.
lychee_db    | Version: '10.5.11-MariaDB-1:10.5.11+maria~focal'  socket: '/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
lychee_db    | 2021-07-11 13:04:23 0 [Note] InnoDB: Buffer pool(s) load completed at 210711 13:04:23
lychee       | **** Make sure the /conf and /uploads folders exist ****
lychee       | **** Create the symbolic link for the /uploads folder ****
lychee       | **** Create the symbolic link for the /sym folder ****
lychee       | **** Create the symbolic link to the old Lychee-Laravel folder ****
lychee       | **** Copy the .env to /conf ****
lychee       | **** Inject .env values ****
lychee       | **** Generate the key (to make sure that cookies cannot be decrypted etc) ****
lychee       |
lychee       | **************************************
lychee       | *     Application In Production!     *
lychee       | **************************************
lychee       |
lychee       | Command Canceled!
lychee       | **** Migrate the database ****
lychee       |
lychee       | Migrating: 2021_05_13_140700_refactor_size_variants
lychee       |
lychee       | In Connection.php line 678:
lychee       |
lychee       |   SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name 'small_w
lychee       |   idth' (SQL: alter table `photos` add `small_width` int unsigned null, add `
lychee       |   small_height` int unsigned null, add `small2x_width` int unsigned null, add
lychee       |    `small2x_height` int unsigned null, add `medium_width` int unsigned null,
lychee       |   add `medium_height` int unsigned null, add `medium2x_width` int unsigned nu
lychee       |   ll, add `medium2x_height` int unsigned null)
lychee       |
lychee       |
lychee       | In Exception.php line 18:
lychee       |
lychee       |   SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name 'small_w
lychee       |   idth'
lychee       |
lychee       |
lychee       | In PDOStatement.php line 112:
lychee       |
lychee       |   SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name 'small_w
lychee       |   idth'
lychee       |
lychee       |
lychee       | **** Check user.css exists and symlink it ****
lychee       | **** Create user and use PUID/PGID ****
lychee       |          User UID :      1000
lychee       |          User GID :      1000
lychee       | **** Set Permissions ****
lychee       | **** Setup complete, starting the server. ****
```

It looks like there are some duplicate columns in the database, or something.
Any help is appreciated, so gallery could be up and running soon :)

thank you

**Comment by kamil4** [comment]
I don't use Docker so I don't understand how it happened (@d7415 maybe able to help) but you somehow ended up with the database in an inconsistent state. It contains columns that shouldn't be there until after the migration runs, which is why the migration fails...

Do you have access to the database? E.g., if you have shell access, run `php artisan db` in the Lychee directory and then invoke the following SQL queries that should give us some clarity on the state of your database:
```
select value from configs where `key` = 'version';
describe photos;
```
Running `php artisan migrate:status` in the shell first may also be revealing...

**Comment by clix10** [comment]
Hello kamil4,

soon after I wrote the post, I was executing  `php artisan migrate`, `php artisan migrate:fresh` and than `php artisan migrate:rollback`, from the docker container shell, like an idiot.

I didn't really know what I was dooing ( obviously :D ) In the end, I ended up with what it looked like a fresh installation of lychee, so I just scraped everything and started from fresh docker containers. 

Setup was fairly straightforward, because I already have predefined folder structure and import files only via simlynk.

But still, I would like to know, what went wrong, so I can avoid it in the future migrations. :)

thank you guys!

**Comment by kamil4** [comment]
Does the start-up log you included at the top come from the very first launch of Lychee after the update or could it be that there was another launch before this one? Basically, I'm speculating that this may not be the first time this particular migration ran. It may have run partially during a preceding launch, created the new columns, but then failed for some reason before finishing. That would leave the database in an inconsistent state (I don't think migrations run as transactions, do they, @ildyria?).

Unfortunately, your subsequent `migrate:fresh` wiped out the database so unless you have a backup, there's no way for us to figure out at this point what originally went wrong and why...

**Comment by d7415** [comment]
It says it's using release, so it should just be migrating from the v4.3.0 to v4.3.4 tag.

**Comment by clix10** [comment]
@kamil4 , you are correct, this is not the log from the very firs start after update. After I already received the error, I restarted it captured a log.

I accessed it via my domain name, but when I pressed migrate, it forwarded to http:\\localhost, and because I run it on a separate server, I lost user interface. I actually waited for few minutes (due to #1033 ) in case it was working in the background than went to the correct domain name and was again greeted with the message that migration is needed.

hope this is at least somewhat helpful :)

**Comment by erkerb4** [comment]
I've run into a similar problem as clix10 _(issues after upgrade),_ and can pick up where @kamil4 responded:

```
sqlite> select value from configs where `key` = 'version';
040301
```

```
root@lychee-d4d6859d9-qvzpv:/var/www/html/Lychee# php artisan migrate:status
+------+-------------------------------------------------------------+-------+
| Ran? | Migration                                                   | Batch |
+------+-------------------------------------------------------------+-------+
| Yes  | 2014_10_12_000000_create_users_table                        | 1     |
| Yes  | 2018_08_03_110935_create_albums_table                       | 1     |
| Yes  | 2018_08_03_110936_create_photos_table                       | 1     |
| Yes  | 2018_08_03_110942_create_configs_table                      | 1     |
| Yes  | 2018_08_03_111324_create_logs_table                         | 1     |
| Yes  | 2018_08_10_134924_move_settings                             | 1     |
| Yes  | 2018_08_15_102039_move_albums                               | 1     |
| Yes  | 2018_08_15_103716_move_photos                               | 1     |
| Yes  | 2018_10_30_135411_sharing                                   | 1     |
| Yes  | 2019_02_21_114356_create_pages_table                        | 1     |
| Yes  | 2019_02_21_114408_create_page_contents_table                | 1     |
| Yes  | 2019_06_21_180451_create_sym_links_table                    | 1     |
| Yes  | 2019_09_28_171753_config_fix                                | 1     |
| Yes  | 2019_09_28_190822_photos_fix                                | 1     |
| Yes  | 2019_10_01_add_livephoto_cols                               | 1     |
| Yes  | 2019_10_02_1400_config_map_display_public                   | 1     |
| Yes  | 2019_10_03_214750_frame_refresh_in_sec                      | 1     |
| Yes  | 2019_10_06_1400_config_map_providers                        | 1     |
| Yes  | 2019_10_06_152017_add_force_32bit_ids                       | 1     |
| Yes  | 2019_10_07_0900_config_map_include_sub_albums               | 1     |
| Yes  | 2019_10_09_233402_config_map_mod                            | 1     |
| Yes  | 2019_10_11_093442_config_check_update_every                 | 1     |
| Yes  | 2019_12_02_2100_config_exiftool                             | 1     |
| Yes  | 2019_12_15_0700_add_share_button_visible_option             | 1     |
| Yes  | 2019_12_15_1000_config_check_update_every_cat_fix           | 1     |
| Yes  | 2019_12_25_0600_config_exiftool_ternary                     | 1     |
| Yes  | 2020_01_018_2300_config_import_via_symlink                  | 1     |
| Yes  | 2020_01_04_1200_config_has_ffmpeg                           | 1     |
| Yes  | 2020_01_26_1200_config_public_sorting                       | 1     |
| Yes  | 2020_01_28_133201_composer_update                           | 1     |
| Yes  | 2020_02_14_0600_location_decoding                           | 1     |
| Yes  | 2020_03_11_124417_increase_length_photo_type                | 1     |
| Yes  | 2020_03_17_200000_unhide_configs                            | 1     |
| Yes  | 2020_04_19_122905_bump_version                              | 1     |
| Yes  | 2020_04_22_155712_bump_version040002                        | 1     |
| Yes  | 2020_04_29_000250_bump_version040003                        | 1     |
| Yes  | 2020_05_12_114228_rss                                       | 1     |
| Yes  | 2020_05_12_161427_bump_version040005                        | 1     |
| Yes  | 2020_05_19_174233_config_prefer_available_xmp_metadata      | 1     |
| Yes  | 2020_05_26_135052_bump_version040006                        | 1     |
| Yes  | 2020_06_04_104605_config_editor_enabled                     | 2     |
| Yes  | 2020_07_11_104605_config_lossless_optimization              | 2     |
| Yes  | 2020_07_11_184605_update_licences                           | 2     |
| Yes  | 2020_07_26_085322_config_swipe_tolerance                    | 2     |
| Yes  | 2020_07_29_132731_config_local_takestamp                    | 2     |
| Yes  | 2020_08_21_123622_add_smart_tag_album_cols                  | 2     |
| Yes  | 2020_10_09_130043_bump_version040007                        | 2     |
| Yes  | 2020_10_15_104504_add_log_max_num_line                      | 3     |
| Yes  | 2020_10_15_161346_sort_image_per_album                      | 3     |
| Yes  | 2020_11_12_183345_config_password_url_param_for_smart_album | 3     |
| Yes  | 2020_11_19_231553_bump_version040008                        | 3     |
| Yes  | 2020_12_12_203153_migrate_admin_user                        | 4     |
| Yes  | 2020_12_12_203831_create_web_authn_tables                   | 4     |
| Yes  | 2020_12_18_162100_bump_version040009                        | 4     |
| Yes  | 2020_12_18_162155_add_nsfw_album                            | 4     |
| Yes  | 2020_12_18_163800_bump_version040010                        | 4     |
| Yes  | 2020_12_24_022307_bump_version040100                        | 4     |
| Yes  | 2020_12_26_153220_nested_set_for_albums                     | 5     |
| Yes  | 2021_01_09_163715_remove_max_min_takestamps                 | 5     |
| Yes  | 2021_01_12_122546_bump_version040200                        | 5     |
| Yes  | 2021_01_18_103729_add_album_cover                           | 5     |
| Yes  | 2021_01_20_113912_bump_version040201                        | 5     |
| Yes  | 2021_01_24_231904_fix-rotation                              | 6     |
| Yes  | 2021_01_27_085903_config_map_display_direction              | 6     |
| Yes  | 2021_01_30_111736_display_takedate                          | 6     |
| Yes  | 2021_02_12_222948_config_upload_processing_limit            | 6     |
| Yes  | 2021_02_13_132245_bump_version040202                        | 6     |
| Yes  | 2021_02_18_232639_config_public_photos_hidden               | 7     |
| Yes  | 2021_03_03_175555_config_remove_image_overlay               | 7     |
| Yes  | 2021_04_17_135924_bump_version040300                        | 7     |
| Yes  | 2021_05_02_174300_add_filesize_raw_col                      | 8     |
| Yes  | 2021_05_12_185726_bump_version040301                        | 8     |
| No   | 2021_05_13_140700_refactor_size_variants                    |       |
| No   | 2021_05_16_171615_bump_version040302                        |       |
| No   | 2021_05_25_160600_post_revert_fixes                         |       |
| No   | 2021_05_31_201000_convert_filesize_to_bigint                |       |
| No   | 2021_06_01_181900_refactor_timestamps_anew                  |       |
| No   | 2021_06_01_182000_bump_version040304                        |       |
| No   | 2021_06_06_151613_fix-takedate                              |       |
| Yes  | 2064_12_25_0000_generate_installed_log                      | 1     |
+------+-------------------------------------------------------------+-------+
```

WebUI provides the following exceptions:

```
PDOException

SQLSTATE[HY000]: General error: 1 duplicate column name: small_width

Facade::__callStatic() in /var/www/html/Lychee/database/migrations/2021_05_13_140700_refactor_size_variants.php (line 41) 
```

```
Doctrine\DBAL\Driver\PDO\ Exception

SQLSTATE[HY000]: General error: 1 duplicate column name: small_width

```

```
Illuminate\Database\QueryException:
SQLSTATE[HY000]: General error: 1 duplicate column name: small_width (SQL: alter table "photos" add column "small_width" integer)
```

**Comment by kamil4** [comment]
I had a feeling this wasn't going to be the last we heard of this problem... :disappointed:

@nagmat84, I know you are busy with your massive PR, but we may need your help here as it's your code that's failing :smiley:.

@erkerb4, thank you for the diagnostics. It confirms that the `2021_05_13_140700_refactor_size_variants` migration has failed in some way. Unfortunately, we don't have the original reason for the failure in the output you provided, because your output is from a rerun, which fails for a different reason than the first attempt.

From looking at this database migration, I can see that @nagmat84 wrapped much of it in a database transaction, so the database shouldn't be _too_ messed up. Ordinarily I would tell you to manually drop the newly added columns and rerun the migration but apparently with SQLite dropping columns may not work. So instead I suggest that you edit the database migration file (it's in `Lychee/database/migrations/2021_05_13_140700_refactor_size_variants.php`), commenting out the first statement in the `up` function:
https://github.com/LycheeOrg/Lychee/blob/8d41c0f928b3ffac8f07523a6b369f3daeee28df/database/migrations/2021_05_13_140700_refactor_size_variants.php#L32-L41
Simply put `//` in front of each of these lines... Then run `php artisan migrate`. It will try to rerun this migration, and I expect it to fail (again), but this time it should generate the diagnostics output that will hopefully allow us to understand what's going wrong and what we need to do to fix it.

**Comment by erkerb4** [comment]
@kamil4 , thank you very much for your response! I've commented out lines 32-41 and re-ran php artisan migrate:

```
**************************************
*     Application In Production!     *
**************************************

 Do you really wish to run this command? (yes/no) [no]:
 > yes

Migrating: 2021_05_13_140700_refactor_size_variants

In SchemaException.php line 86:

  There is no column with name 'small' on table 'photos'.
```

There is no response when I try to display all the data in photos:

```
sqlite> SELECT * FROM photos;
sqlite>
```

**Comment by kamil4** [comment]
That is _very_ bad. :sob:

If the photos table is empty then you've basically lost all your data and will have to start from scratch. _How_ that might have happened I have no idea. Out of curiosity, are any of the other tables empty as well? `albums`, `users`, etc?

Do you have a backup of your sqlite database by any chance? That's your only hope to recover anything at this point... :disappointed:

**Comment by erkerb4** [comment]
I've got to see how current my backup is.. I may choose to start fresh, and maybe use an external db instance this time around. 

It seems like other tables are in good condition: albums, configs, logs, users have valid content.

**Comment by kamil4** [comment]
Could it be that you ran out of disk space? My understanding is that in many cases altering tables in SQLite effectively means creating a new table and populating it with the data from the old one. Do you know how large your photos table was? How many photos did you have in your Lychee instance?

**Comment by erkerb4** [comment]
I do not have a definitive answer, however It is possible that I may have run out of disk space on the /conf partition where the DB is stored. It is a 5GB partition.

I think I had about 6k photos in the Lychee instance. 

I appreciate your help, and responses!

**Comment by derikshen** [comment]
Hi similar issue don't know if I should open new issue; 
I appreciate any assistance you can provide :)


Migrating: 2021_05_02_174300_add_filesize_raw_col

In Connection.php line 678:

  SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name 'filesize' (SQL: alter table `photos` add `filesize` bigint unsigned not null default '0' after `size`)


In Exception.php line 18:

  SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name 'filesize'


In PDOStatement.php line 112:

  SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name 'filesize'
